ext/url/php_url.c

```diff
@@ -0,0 +1,423 @@
+		Z_PARAM_STR_OR_NULL(scheme)
+		Z_PARAM_STR_OR_NULL(host)
+		Z_PARAM_LONG_OR_NULL(port, port_is_null)
+		Z_PARAM_STR_OR_NULL(user)
+		Z_PARAM_STR_OR_NULL(password)
+		Z_PARAM_STR_OR_NULL(path)
+		Z_PARAM_STR_OR_NULL(query)
+		Z_PARAM_STR_OR_NULL(fragment)
+	ZEND_PARSE_PARAMETERS_END();
+
+	UPDATE_STR_PROPERTY(execute_data, ZSTR_KNOWN(ZEND_STR_SCHEME), scheme);
+	UPDATE_STR_PROPERTY(execute_data, ZSTR_KNOWN(ZEND_STR_HOST), host);
+	UPDATE_LONG_PROPERTY(execute_data, ZSTR_KNOWN(ZEND_STR_PORT), port, port_is_null);
+	UPDATE_STR_PROPERTY(execute_data, ZSTR_KNOWN(ZEND_STR_USER), user);
+	UPDATE_STR_PROPERTY(execute_data, ZSTR_KNOWN(ZEND_STR_PASSWORD), password);
+	UPDATE_STR_PROPERTY(execute_data, ZSTR_KNOWN(ZEND_STR_PATH), path);
+	UPDATE_STR_PROPERTY(execute_data, ZSTR_KNOWN(ZEND_STR_QUERY), query);
+	UPDATE_STR_PROPERTY(execute_data, ZSTR_KNOWN(ZEND_STR_FRAGMENT), fragment);
+}
+
+PHP_METHOD(Url_Url, __toString)
+{
+	ZEND_PARSE_PARAMETERS_NONE();
+
+	// TODO Not implemented yet
+	RETURN_EMPTY_STRING();
+}
+
+PHP_METHOD(Url_UrlParser, parseUrlComponent)
+{
+	zend_string *url;
+	zend_object *component;
+
+	ZEND_PARSE_PARAMETERS_START(2, 2)
+		Z_PARAM_STR(url)
+		Z_PARAM_OBJ_OF_CLASS(component, url_component_ce)
+	ZEND_PARSE_PARAMETERS_END();
+
+	// TODO Not implemented yet
+	RETURN_EMPTY_STRING();
+}
+
+PHP_METHOD(Url_UrlParser, parseUrl)
+{
+	zend_string *url;
+
+	ZEND_PARSE_PARAMETERS_START(1, 1)
+		Z_PARAM_PATH_STR(url)
+	ZEND_PARSE_PARAMETERS_END();
+
+	parse_url_to_object(url, return_value);
+}
+
+PHP_METHOD(Url_UrlParser, parseUrlArray)
+{
+	zend_string *url;
+
+	ZEND_PARSE_PARAMETERS_START(1, 1)
+		Z_PARAM_PATH_STR(url)
+	ZEND_PARSE_PARAMETERS_END();
+
+	parse_url_to_array(url, return_value);
+}
+
+void url_register_symbols(void)
+{
+	url_ce = register_class_Url_Url(zend_ce_stringable);
+	url_component_ce = register_class_Url_UrlComponent();
+	url_parser_ce = register_class_Url_UrlParser();
+}
+
+PHP_MINIT_FUNCTION(url)
+{
+	url_register_symbols();
+
+	return SUCCESS;
+}
+
+PHP_MSHUTDOWN_FUNCTION(url)
+{
+	return SUCCESS;
+}
+
+PHP_RINIT_FUNCTION(url)
+{
+#if defined(COMPILE_DL_URL) && defined(ZTS)
+	ZEND_TSRMLS_CACHE_UPDATE();
+#endif
+	URL_G(parser) = init_parser();
+	URL_G(urls) = 0;
+
+	return URL_G(parser) == NULL ? FAILURE : SUCCESS;
+}
+
+PHP_RSHUTDOWN_FUNCTION(url)
+{
+	destroy_parser(URL_G(parser));
+
+	return SUCCESS;
+}
+
+zend_module_entry url_module_entry = {
+	STANDARD_MODULE_HEADER,
+	"url",                             /* Extension name */
+	NULL,                           /* zend_function_entry */
+	PHP_MINIT(url),         /* PHP_MINIT - Module initialization */
+	PHP_MSHUTDOWN(url),   /* PHP_MSHUTDOWN - Module shutdown */
+	PHP_RINIT(url),         /* PHP_RINIT - Request initialization */
+	PHP_RSHUTDOWN(url),   /* PHP_RSHUTDOWN - Request shutdown */
+	NULL,                            /* PHP_MINFO - Module info */
+	PHP_VERSION,                      /* Version */
+	STANDARD_MODULE_PROPERTIES
```

Comment:
The reason ZTS crashes is because you never specified the module globals in this structure, which means the TSRM doesn't allocate enough memory for this.